File: src/scripts/tools/adapter_errors.py
#!/usr/bin/env python3
"""
Shared error types for tool adapters.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Optional


class AdapterErrorType(str, Enum):
    """Categorized adapter errors."""
    AUTH_FAILURE = "auth_failure"
    NOT_FOUND = "not_found"
    RATE_LIMIT = "rate_limit"
    TIMEOUT = "timeout"
    NETWORK = "network"
    VALIDATION = "validation"
    UNKNOWN = "unknown"


@dataclass
class AdapterError:
    """Structured adapter error."""
    adapter: str
    action: str
    error_type: AdapterErrorType
    message: str
    status_code: Optional[int] = None
    retry_after: Optional[int] = None

    @property
    def is_retryable(self) -> bool:
        """Whether this error can be retried."""
        return self.error_type in (
            AdapterErrorType.RATE_LIMIT,
            AdapterErrorType.TIMEOUT,
            AdapterErrorType.NETWORK,
        )


def classify_http_error(adapter: str, action: str, status_code: int, message: str) -> AdapterError:
    """Classify an HTTP error into a structured AdapterError."""
    if status_code == 401 or status_code == 403:
        error_type = AdapterErrorType.AUTH_FAILURE
    elif status_code == 404:
        error_type = AdapterErrorType.NOT_FOUND
    elif status_code == 429:
        error_type = AdapterErrorType.RATE_LIMIT
    elif status_code >= 500:
        error_type = AdapterErrorType.NETWORK
    else:
        error_type = AdapterErrorType.UNKNOWN

    return AdapterError(
        adapter=adapter,
        action=action,
        error_type=error_type,
        message=message,
        status_code=status_code,
    )

File: src/scripts/tools/adapter_errors.ts
/**
 * Shared error types for tool adapters.
 *
 * TypeScript twin of `src/scripts/tools/adapter_errors.py` (ADR-096 —
 * Python→TS migration, Phase 1 / tools adapter cluster). Public API mirrors
 * the Python module exactly (snake_case kept deliberately — fidelity over TS
 * idiom): the `AdapterErrorType` string enum, the `AdapterError` value shape
 * with its `is_retryable` property, and `classify_http_error`'s status-code →
 * error-type mapping. No behaviour changes.
 */

/** Categorized adapter errors (mirrors the `AdapterErrorType` str-Enum). */
export enum AdapterErrorType {
    AUTH_FAILURE = 'auth_failure',
    NOT_FOUND = 'not_found',
    RATE_LIMIT = 'rate_limit',
    TIMEOUT = 'timeout',
    NETWORK = 'network',
    VALIDATION = 'validation',
    UNKNOWN = 'unknown',
}

/**
 * Structured adapter error (mirrors the `AdapterError` dataclass).
 *
 * Construct via `makeAdapterError` to keep the same field defaults as the
 * Python dataclass (`status_code` / `retry_after` default to null).
 */
export interface AdapterError {
    adapter: string;
    action: string;
    error_type: AdapterErrorType;
    message: string;
    status_code: number | null;
    retry_after: number | null;
}

/** Build an `AdapterError` with Python dataclass defaults. */
export function makeAdapterError(args: {
    adapter: string;
    action: string;
    error_type: AdapterErrorType;
    message: string;
    status_code?: number | null;
    retry_after?: number | null;
}): AdapterError {
    return {
        adapter: args.adapter,
        action: args.action,
        error_type: args.error_type,
        message: args.message,
        status_code: args.status_code ?? null,
        retry_after: args.retry_after ?? null,
    };
}

/** Whether this error can be retried (mirrors the `is_retryable` property). */
export function is_retryable(error: AdapterError): boolean {
    return (
        error.error_type === AdapterErrorType.RATE_LIMIT ||
        error.error_type === AdapterErrorType.TIMEOUT ||
        error.error_type === AdapterErrorType.NETWORK
    );
}

/** Classify an HTTP error into a structured AdapterError. */
export function classify_http_error(
    adapter: string,
    action: string,
    status_code: number,
    message: string,
): AdapterError {
    let error_type: AdapterErrorType;
    if (status_code === 401 || status_code === 403) {
        error_type = AdapterErrorType.AUTH_FAILURE;
    } else if (status_code === 404) {
        error_type = AdapterErrorType.NOT_FOUND;
    } else if (status_code === 429) {
        error_type = AdapterErrorType.RATE_LIMIT;
    } else if (status_code >= 500) {
        error_type = AdapterErrorType.NETWORK;
    } else {
        error_type = AdapterErrorType.UNKNOWN;
    }

    return makeAdapterError({
        adapter,
        action,
        error_type,
        message,
        status_code,
    });
}
